src/Callable.cpp

```diff
@@ -0,0 +1,178 @@
+        for (size_t i = 0; i < argc; i++) {
+            if (actual_qcci[i] != expected_qcci[i]) {
+                do_check_fail(i, argc, "calling");
+            }
+        }
+    } else {
+        do_check_fail(-1, argc, "calling");
+    }
+}
+
+void Callable::check_fcci(size_t argc, const FullCallCheckInfo *actual_fcci) const {
+    // Lazily create full_call_check_info upon the first call to make_std_function().
+    if (contents->full_call_check_info.empty()) {
+        contents->full_call_check_info.reserve(contents->jit_cache.arguments.size());
+        for (const Argument &a : contents->jit_cache.arguments) {
+            const auto fcci = a.is_scalar() ? Callable::make_scalar_fcci(a.type) : Callable::make_buffer_fcci(a.type, a.dimensions);
+            contents->full_call_check_info.push_back(fcci);
+        }
+    }
+
+    const size_t required_arg_count = contents->full_call_check_info.size();
+    if (argc == required_arg_count) {
+        const FullCallCheckInfo *expected_fcci = contents->full_call_check_info.data();
+        for (size_t i = 0; i < argc; i++) {
+            if (!Callable::is_compatible_fcci(actual_fcci[i], expected_fcci[i])) {
+                do_check_fail(i, argc, "defining");
+            }
+        }
+    } else {
+        do_check_fail(-1, argc, "defining");
+    }
+}
+
+// Entry point used from the std::function<> variant; we can skip the check_qcci() stuff
+// since we verified the signature when we created the std::function, so incorrect types or counts
+// should be impossible.
+/*static*/ int Callable::call_argv_fast(size_t argc, const void *const *argv) const {
+    // Callable should enforce these, so we can use assert() instead of internal_assert()
+    assert(contents->jit_cache.jit_target.has_feature(Target::UserContext));
+    assert(contents->jit_cache.arguments[0].name == "__user_context");
+
+    JITUserContext *context = *(JITUserContext **)const_cast<void *>(argv[0]);
+    assert(context != nullptr);
+
+    JITFuncCallContext jit_call_context(context, contents->saved_jit_handlers);
+
+    // debug(2) << "Calling jitted function\n";
+    int exit_status = contents->jit_cache.call_jit_code(contents->jit_cache.jit_target, argv);
+    // debug(2) << "Back from jitted function. Exit status was " << exit_status << "\n";
+
+    // If we're profiling, report runtimes and reset profiler stats.
+    contents->jit_cache.finish_profiling(context);
+
+    // Don't call jit_call_context.finalize(): if no error handler was installed,
+    // it will call halide_error_handler for us, which is fine for realize()
+    // and friends because there is no other way to report an error. For this code
+    // path, though, we just return the error code (if any); if a custom error
+    // handler was installed, it presumably will get the first shot at handling it,
+    // and if not, the caller must handle it.
+    //
+    // No: jit_call_context.finalize(exit_status);
```

Comment:
Not sure what these two lines are about. They seem redundant.

Reply:
Scalpel left in patient, removed.